### PR TITLE
driver/serial: Convert CR to LF in driver

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary

Enable the behavior by default for console, but configurable by termios.

Binary size:

Before:
   text    data     bss     dec     hex filename
 326460     409    8164  335033   51cb9 nuttx/nuttx

After:
   text    data     bss     dec     hex filename
 326478     409    8164  335051   51ccb nuttx/nuttx

I'm trying to let nsh get rid of the stdio by these steps:
1. Convert CR to LF in serial input (follow linux, this patch)
2. Support O_TEXT and O_BINARY to convert CR2LF in VFS layer
3. Remove CONFIG_EOL_IS_XXX confgis
4. Echo character in serial driver instead of nsh

## Impact
serial driver
## Testing
QEMU

### CODE REVIEW

**Review by acassis — APPROVED** [review]

**Comment by acassis** [comment]
It increased more than 50 bytes, but since it remove those CONFIG_EOL_IS_XXX configs, this is a fair price to pay!

**Comment by acassis** [comment]
Please check these issues:
```
 serial/serial.c: In function 'uart_ioctl':
Error: serial/serial.c:1397:46: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
 1397 |               FAR struct termios *termiosp = (FAR struct termios *)arg;
      |                                              ^
Error: serial/serial.c:1419:46: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
 1419 |               FAR struct termios *termiosp = (FAR struct termios *)arg;
      |    
``

File: drivers/serial/serial.c [hunk 1]
```diff
@@ -1647,21 +1653,25 @@ int uart_register(FAR const char *path, FAR uart_dev_t *dev)
...
 
       /* Enable \n -> \r\n translation for the console */
 
       dev->tc_oflag = OPOST | ONLCR;
-    }
 #endif
 
+      /* Convert CR to LF by default for console */
+
+      dev->tc_iflag |= ICRNL;
```
Comment (xiaoxiang781216) [inline drivers/serial/serial.c:1672]: let's follow how tc_oflag get processed specially for console in case of CONFIG_SERIAL_TERMIOS isn't enable.

File: include/nuttx/serial/serial.h [hunk 2]
```diff
@@ -283,10 +283,10 @@ struct uart_dev_s
   pid_t                pid;          /* Thread PID to receive signals (-1 if none) */
 #endif
 
-#ifdef CONFIG_SERIAL_TERMIOS
   /* Terminal control flags */
 
   tcflag_t             tc_iflag;     /* Input modes */
+#ifdef CONFIG_SERIAL_TERMIOS
```
Comment (xiaoxiang781216) [inline include/nuttx/serial/serial.h:289]: revert the change, let's check isconsole to do the input conversion if CONFIG_SERIAL_TERMIOS isn't enabled.

Comment (no1wudi) [reply include/nuttx/serial/serial.h:289]: That means if user want to disable the input conversion, them must enable TERMIOS to do that?

Comment (xiaoxiang781216) [reply include/nuttx/serial/serial.h:289]: isconsole control the default behaviour:

1. If isconsole equals true, the new line conversion for both direction and the check of Ctrl+C check will happen.
2. If isconsole equals false, no interception happens in serial driver level.

CONFIG_SERIAL_TERMIOS enable the user change the default behaviour at the runtime.

Comment (no1wudi) [reply include/nuttx/serial/serial.h]: Get it

**Comment by no1wudi** [comment]
> Please check these issues:
> 
> ```
>  serial/serial.c: In function 'uart_ioctl':
> Error: serial/serial.c:1397:46: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
>  1397 |               FAR struct termios *termiosp = (FAR struct termios *)arg;
>       |                                              ^
> Error: serial/serial.c:1419:46: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
>  1419 |               FAR struct termios *termiosp = (FAR struct termios *)arg;
>       |    
> ``
> ```

Fixed

**Review by xiaoxiang781216 — APPROVED** [review]
